Set default qengine to QNNPACK on ARM for quantization tests

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83125
* #83115
* __->__ #83097
* #83089

The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

soulitzer committed:
Set default qengine to QNNPACK on ARM for quantization tests

[ghstack-poisoned]

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


There is no FBGEMM or MKLDNN on ARM so these tests were previously skipped (and QNNPACK is disabled by default for some reason). Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_docs.py:
```diff
@@ -14,6 +14,16 @@
 
+@contextlib.contextmanager
+def use_qnnpack_qengine():
+    prev = torch._C._get_qengine()
+    torch._C._set_qengine(2)
```
there is a public api for these I think, something like the following:
read: torch.backends.quantized.engine
write: torch.backends.quantized.engine = "qnnpack" 

you can also use https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_op.py#L521 to overwrite the quantized engine

soulitzer replied:
Cool, thanks!

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

jerryzh168 reviewed (approved):
lg, please make sure ci passes before landing

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]

albanD commented on test/quantization/core/test_docs.py:
```diff
@@ -108,21 +109,23 @@ def _test_code(self, code, global_inputs=None):
         unique_identifier = "PTDQ API Example::"
-        code = self._get_code(path_from_pytorch, unique_identifier)
-        self._test_code(code)
+        with override_quantized_engine("qnnpack") if IS_ARM64 else contextlib.nullcontext():
```
If we do this for every test in here, should we do it in the `setUp` function instead?

albanD reviewed (approved):
Sounds ok.
Small nit only

soulitzer committed:
Update on "Set default qengine to QNNPACK on ARM for quantization tests"


The issue is that there is no FBGEMM or MKLDNN on ARM and QNNPACK is disabled by default for some reason. Not sure if this PR is the right approach, alternatives are probably:
1) keep these tests skipped
2) not disable QNNPACK by default on ARM
3) ?

[ghstack-poisoned]
